hare4/hare.go

```diff
@@ -283,6 +308,135 @@ func (h *Hare) Running() int {
 	return len(h.sessions)
 }
 
+// fetchFull will fetch the full list of proposal IDs from the provided peer.
+func (h *Hare) fetchFull(ctx context.Context, peer p2p.Peer, msgId types.Hash32) (
+	[]types.ProposalID, error,
+) {
+	ctx, cancel := context.WithTimeout(ctx, fetchFullTimeout)
+	defer cancel()
+
+	requestCompactCounter.Inc()
+	req := &CompactIdRequest{MsgId: msgId}
+	reqBytes := codec.MustEncode(req)
+	resp := &CompactIdResponse{}
+	cb := func(ctx context.Context, rw io.ReadWriter) error {
+		respLen, _, err := codec.DecodeLen(rw)
+		if err != nil {
+			return fmt.Errorf("decode length: %w", err)
+		}
+		if respLen >= MAX_EXCHANGE_SIZE {
+			return errResponseTooBig
+		}
+		buff := make([]byte, respLen)
+		_, err = io.ReadFull(rw, buff)
+		if err != nil {
+			return fmt.Errorf("read response buffer: %w", err)
+		}
+		err = codec.Decode(buff, resp)
```

Comment:
Can't we use `codec.DecodeFrom(rw, resp)`?
We can then check return value (byte count) and see if it's the same as `respLen`.
This way, GC pressure will be reduced somewhat as we will not need the temporary buffer.